cou/steps/plan.py

```diff
@@ -435,39 +447,84 @@ def _generate_control_plane_plan(
     return [principal_upgrade_plan, subordinate_upgrade_plan]
 
 
-async def _generate_data_plane_plan(
-    target: OpenStackRelease,
-    analysis_result: Analysis,
+def _separate_hypervisors_apps(
+    apps_data_plane: list[OpenStackApplication],
+) -> tuple[list[OpenStackApplication], list[OpenStackApplication]]:
+    """Separate what is considered hypervisors apps from non-hypervisors apps.
+
+    :param apps_data_plane: Applications from data-plane
+    :type apps_data_plane: list[OpenStackApplication]
+    :return: Tuple containing two lists of hypervisors and non-hypervisors apps
+    :rtype: tuple[list[OpenStackApplication], list[OpenStackApplication]]
+    """
+    hypervisor_apps = []
+    non_hypervisors_apps = []
+    for app in apps_data_plane:
+        if app.charm == "ceph-osd" or app.is_subordinate:
+            non_hypervisors_apps.append(app)
+            continue
+
+        hypervisor_apps.append(app)
```

Comment:
not-blocking: In this case I think it cleaner to use if else.
```suggestion
        if app.charm == "ceph-osd" or app.is_subordinate:
            non_hypervisors_apps.append(app)
        else:
            hypervisor_apps.append(app)
```